### PR TITLE
mmap and vmsplice support

### LINKED ISSUE #200: vmsplice() causes LKL to crash (segfault)
I tried to use vmsplice() (through hijack) on LKL but my program crashed upon the 1st invocation of vmsplice(). I've tracked down the crash to be due to a null "current->mm" when __get_user_pages_unlocked() in mm/nommu.c was called. "current" is the currently executing (kernel) thread->task and mm is of struct mm_struct * inside struct task_struct. It looks like "active_mm" is fine:
(gdb) p ((struct task_struct *)0x7f942b8ea300)->active_mm
$7 = (struct mm_struct *) 0x7f94327dec60 <init_mm>

Am I hitting a possible kernel bug, or this is an issue with LKL?

 I've captured the stack trace through gdb as below:
Program received signal SIGSEGV, Segmentation fault.
__down_read (sem=0x58) at kernel/locking/rwsem-spinlock.c:136
136     kernel/locking/rwsem-spinlock.c: No such file or directory.
(gdb) bt
#0  __down_read (sem=0x58) at kernel/locking/rwsem-spinlock.c:136
#1  0x00007f7b4eb05ac9 in down_read (sem=0x0) at kernel/locking/rwsem.c:24
#2  0x00007f7b4e7b1a99 in __get_user_pages_unlocked (tsk=0x1, mm=0x58, start=35995648, nr_pages=1,
    write=0, force=0, pages=0x7f7b4f095c90, gup_flags=0) at mm/nommu.c:192
#3  0x00007f7b4e7b1b1f in get_user_pages_unlocked (start=1, nr_pages=1, write=35995648, force=0, pages=0x0)
    at mm/nommu.c:203
#4  0x00007f7b00000000 in ?? ()
#5  0x00007f7b4f095bf0 in ?? ()
#6  0x00007f7b4e7aba01 in get_user_pages_fast (start=0, nr_pages=0, write=35995648, pages=0x0)
    at mm/util.c:286
#7  0x00007f7b4e7e2def in get_iovec_page_array (aligned=<optimized out>, pipe_buffers=<optimized out>,
    partial=<optimized out>, pages=<optimized out>, nr_vecs=<optimized out>, iov=<optimized out>)
    at fs/splice.c:1493
#8  vmsplice_to_pipe (file=0x0, iov=0x58, nr_segs=88, flags=0) at fs/splice.c:1613
#9  0x00007f7b47a5c580 in ?? ()
#10 0x00007f7b4f095c60 in ?? ()
#11 0x0000000000405c50 in iov ()
#12 0x00007f7b4f095c50 in ?? ()
#13 0x00007f7b4f095c90 in ?? ()
#14 0x0000000000000000 in ?? ()
(gdb)

### PR DESCRIPTION
**Pull request by ghost**
This PR adds support for mmap and vmsplice (as long as the buffer used in vmsplice is allocated with mmap).

It fixes #200.

---

This change is [Reviewable](https://reviewable.io/reviews/lkl/linux/204)

### CODE REVIEW

**Comment by hkchu** [comment]
I tried the fix but the kernel crashed when coming up: (Do i need a "/init" script?)
...
[    1.770000] Btrfs loaded
[    1.770000] Warning: unable to open an initial console.
[    1.770000] This architecture does not have kernel memory protection.
[    1.770000] Failed to execute /init (error -13)
[    1.770000] Kernel panic - not syncing: No working init found.  Try passing init= option to kernel. See Linux Documentation/init.txt for guidance.
[    1.770000] ---[ end Kernel panic - not syncing: No working init found.  Try passing init= option to kernel. See Linux Documentation/init.txt for guidance.
./lkl-hijack.sh: line 22: 317823 Segmentation fault      (core dumped) LD_PRELOAD=liblkl-hijack.so $*

**Comment by opurdila** [comment]
That is weird, it passed the tests on my machine and on CircleCI. Maybe try a clean build? If it still fails, could you describe what are you using to cause the crash?

**Comment by opurdila** [comment]
@hkchu please check my latest commits. It allows me to run the following simple vmsplice program:

```
#define _GNU_SOURCE
#include <fcntl.h>
#include <sys/uio.h>
#include <stdio.h>
#include <unistd.h>
#include <sys/mman.h>

#define BUF_SIZE        8192

int main(void)
{
        int pfd[2];
        struct iovec iov;


        iov.iov_base = mmap(NULL, BUF_SIZE, PROT_READ|PROT_WRITE,
                             MAP_ANONYMOUS|MAP_PRIVATE, -1, 0);
        iov.iov_len = BUF_SIZE;

        printf("%s: %p\n", __func__, iov.iov_base);

        if (pipe(pfd))
                perror("pipe");
        if (vmsplice(pfd[1], &iov, 1, 0) != 8192)
                perror("vmsplice");
}
```

**Comment by hkchu** [comment]
I don't know what i've missed but I can't get your latest commit to boot - the kernel got stuck at
lkl_ops->sem_down(init_sem);

inside lkl_start_kernel() because lkl_run_init() is never called. Unfortunately I haven't got a handle on how this init stuff works. I'm running the test through hijack but I assume this problem has nothing to do with hijack.

**Comment by opurdila** [comment]
Did you try a clean build? Do you have the branch available for me to take a look?

Last resort we can try creating a fake PR from your end so that Circle CI runs and we can both run / debug in the same environment.

**Comment by hkchu** [comment]
What's the proper step for a "clean build"? I usually did "make clean" under tools/lkl.
But this time I also went up to the top of kernel tree and did "make mrproper" there just to be sure.
"fs_setup()" was not being called either. Who is supposed to call fs_setup()?

**Comment by opurdila** [comment]
Yeah, make mrproper at the top level should do the trick. fs_setup should be called from do_init_calls.

Are you using any special compilers? I run into issues like these when the compiler / linker did not properly implemented weak symbols handling (e.g. mingw).

**Comment by hkchu** [comment]
I'm using a child that is direct descend of your init-exec branch. I just searched the whole source tree but could not find any symbol like "do_init_calls". I'm using gnu from my ubuntu desktop.
(3024): gcc --version
gcc (Ubuntu 4.8.4-2ubuntu1~14.04.3) 4.8.4
Copyright (C) 2013 Free Software Foundation, Inc.
...

Is there a workaround for this problem, like calling fs_setup directly from somewhere?

**Comment by opurdila** [comment]
Sorry, the correct function name is do_initcalls. But if the fs_setup is not called probably other things are messed up as well. I've just did a git checkout lkl-opurdila/init-exec and everything seems fine on my end...

**Comment by hkchu** [comment]
start_kernel()->bootmem_init() and the latter never returns... (perhaps that thread dies.)

**Comment by hkchu** [comment]
mem_size seems to be crap.

**Comment by hkchu** [comment]
no, mem_size is fine. The problem is bootmem_init() calling lkl_ops->mem_alloc() got stuck and never returned from malloc() call. Don't know why.

File: tools/lkl/lib/hijack/hijack.c [hunk 1]
```diff
@@ -338,3 +341,14 @@ int eventfd_write(int fd, uint64_t value)
 	return lkl_sys_write(fd, (void *) &value,
 			     sizeof(value)) != sizeof(value) ? -1 : 0;
 }
+
+HOST_CALL(mmap)
+void *mmap(void *addr, size_t length, int prot, int flags, int fd, off_t offset)
+{
+	CHECK_HOST_CALL(mmap);
+
+	if (addr != NULL || flags != (MAP_ANONYMOUS|MAP_PRIVATE) ||
```
Comment (liuyuan10) [inline tools/lkl/lib/hijack/hijack.c]: The libc we are using calls mmap in malloc which get hijacked with this change. Maybe you want to use fd != -2

Comment (hkchu) [inline tools/lkl/lib/hijack/hijack.c]: Good catch! It looks like lkl_host_ops->mem_alloc got hijacked - one of the downside of using the hijack library. More code may get hijacked than one expects...

**Comment by hkchu** [comment]
I've verified the fix works as long as one uses mmap() to allocate user buffers. I've only tested the TX side vmsplice(mmaped buf->pipe[1]) followed by splice(pipe[0]->socketfd), as the RX side vmsplice() uses copy to move data from a pipe to the user buffer so there is no point in using it.

We'll also need the change of the fd check to -2 to avoid other implicit mmap() calls being hijacked, unless there is a better solution. Ultimately one should change the app to call LKL directly instead of relying on hijack I guess.

**Comment by hkchu** [comment]
*[tools/lkl/lib/hijack/hijack.c, line 350 \[r2\]](https://reviewable.io:443/reviews/lkl/linux/204#-KPPLrlprnIpwhRq75AD-r2-350:-KQ3q2ZgANnCIRVWYKCw:bfbbngn) ([raw file](https://github.com/lkl/linux/blob/f6482f150a3c3164d0d93c83eba08a6eac28256b/tools/lkl/lib/hijack/hijack.c#L350)):*
<details><summary><i>Previously, hkchu (Jerry Chu) wrote…</i></summary>
> Good catch! It looks like lkl_host_ops->mem_alloc got hijacked - one of the downside of using the hijack library. More code may get hijacked than one expects...
</details>
I don't quite understand the magic of lkl_run_init(). Will __bprm_mm_init() gets us a mm_struct that accurately describes the kernel address space? In any case the fix plus the change suggested by Yuan works for my test program now. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/lkl/linux/204)*

**Comment by opurdila** [comment]
Good catch indeed, thanks @liuyan10 . Using -2 seems like a good workaround for now, I hope to upload a new version tonight.

**Comment by opurdila** [comment]
> I don't quite understand the magic of lkl_run_init(). Will __bprm_mm_init() gets us a mm_struct that accurately describes the kernel address space?

It will create an empty mm_struct for init's userspace where all syscalls run. Then the mmap call will add some buffer to the address space for vmsplice to find.

**Comment by hkchu** [comment]
:lgtm:

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/lkl/linux/204#-:-KQ8po9wN_OE2rHJoPkV:bnfp4nl)*

**Comment by hkchu** [comment]
Reviewed 3 of 3 files at r1, 3 of 5 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/lkl/linux/204)*

**Pull request merged**